plugins.d/Lets_Encrypt/dns_01.py

```diff
@@ -3,73 +3,197 @@
+    '''Simple subprocess wrapper for running commands'''
+    if env is None:
+        env = {}
+    proc = subprocess.run(command, env=env, capture_output=True, text=True)
+    if proc.returncode != 0:
+        com = ' '.join(command)
+        return (proc.returncode,
+                f"Something went wrong when running '{com}':\n\n{proc.stderr}")
+    else:
+        return 0, 'success'
+
+
+def apt_install(pkgs: list[str]) -> tuple[int, str]:
+    """Takes a list of package names, updates apt and installs packages,
+    returns tuple(exit_code, message)
+    """
+    env = {'DEBIAN_FRONTEND': 'noninteractive'}
+    for command in [['apt-get', 'update'],
+                    ['apt-get', 'install', *pkgs, '-y']]:
+        exit_code, string = run_command(command, env=env)
+        if exit_code != 0:
+            return exit_code, string
+    return exit_code, string
+
+
+def check_pkg(pkg: str) -> bool:
+    """Takes a package name and returns True if installed, otherwise False"""
+    p = subprocess.run(['apt-cache', 'policy', pkg],
+                       capture_output=True, text=True)
+    for line in p.stdout.split('\n'):
+        line = line.strip()
+        if line.startswith('Installed'):
+            if not line.endswith('(none)'):
+                return True # package installed
+            else:
+                return False # package probably available but not installed
+    return False # package uninstallable
+
+
+def initial_setup() -> None:
+    '''Check lexicon and deps are installed and ready to go
+
+    Returns a tuple of exit code (0 = success) and message
+    '''
+    msg_start = 'lexicon tool is required for dns-01 challenge, '
+    msg_mid = ''
+    msg_end = '\n\nDo you wish to continue?'
+    msg = ''
+    remove_lexicon = False
+    install_venv = False
 
-def get_providers():
     lexicon_bin = which('lexicon')
+    venv = '/usr/local/src/venv/lexicon'
+    lexicon_venv_bin = join(venv, 'bin/lexicon')
+    lexicon_venv_bin_syml = '/usr/local/bin/lexicon'
     if not lexicon_bin:
-        ret = console.yesno(
-            'lexicon tool is required to use dns-01 challenge, '
-            'however it is not found on your system.\n\n'
-            'Do you wish to install it now?',
-            autosize=True
-        )
+        # lexicon not found - offer to install via venv
+        install_venv = True
+        msg_mid = "however it is not found on your system, so installing."
+    elif (exists(lexicon_venv_bin) and exists(lexicon_venv_bin_syml)
+            and realpath(lexicon_bin) == lexicon_venv_bin):
+        # lexicon venv found - no message required
+        msg = None
+    elif lexicon_bin == '/usr/bin/lexicon' and check_pkg('lexicon'):
+        # lexicon pkg installed - offer to remove and install via venv
+        install_venv = True
+        remove_lexicon = True
+        msg_mid = ("lexicon deb install noted, but we need a newer version"
+                   " removing deb and install via venv.")
```

Comment:
Not entirely sure about this. Imo if we're installing our own copy of lexicon where it won't be used by the system, we probably shouldn't be modifying/removing system packages if possible.

Reply:
Whilst I want to argue (cause it would be quicker/easier for me) you're right...

Part of the reason why I did things the way that I did was because I found that it was tricky to enable the venv via subprocess (unless using `shell=True` - which I wanted to avoid). I guess it could be done via `env=...` but that would require me parsing the venv `activate` and ensuring that the env is set up right.

So I'll make a wrapper script for our lexicon venv (actually I've already done it, just haven't pushed yet).

Reply:
I've implemented a wrapper script; `turnkey-lexicon`...